pkg/proxy/endpoints.go

```diff
@@ -132,15 +137,29 @@ func (ect *EndpointChangeTracker) Update(previous, current *v1.Endpoints) bool {
 		change = &endpointsChange{}
 		change.previous = ect.endpointsToEndpointsMap(previous)
 		ect.items[namespacedName] = change
+		if t := getLastChangeTriggerTime(endpoints); !t.IsZero() {
+			ect.lastChangeTriggerTimes[namespacedName] = t
+		}
 	}
 	change.current = ect.endpointsToEndpointsMap(current)
 	// if change.previous equal to change.current, it means no change
 	if reflect.DeepEqual(change.previous, change.current) {
 		delete(ect.items, namespacedName)
+		delete(ect.lastChangeTriggerTimes, namespacedName)
```

Comment:
What happen if there are 2 consecutive events, one event has different endpoints, the other event has the same endpoints? Will the 2nd one erase the timestamp of the first one? Causing the 1st event was not measured? Will this happen?

Reply:
IIUC, the situation you described is something like this:

T0: proxier.Sync()
T1: proxier observes Endpoints E1 change, E1.EndpointsLastChangeTriggerTime = t0
T2: proxier observes Endpoints E1 change, E1.EndpointsLastChangeTriggerTime = t1 (t1>=t0)
T3: proxier.Sync()

In such case the implementation will ignore the second timestamp and use t0 (which is guaranteed to be <= t1) to measure the latency. 

Let me know if it makes sense.

Reply:
I agree with reporting multiple triggerTimes for the same endpoint per sync. It makes sense and should be relatively cheap. I updated the code to do that.

Please note that regarding opposite events, that combined reduce to no-change, I left the logic as was, i.e. I'm not exporting the metric for such events.
The definition of Network Programming Latency SLI is (more or less): duration between a time of a change and a time when the network was programmed to incorporate that change. If there are events A and B that together result in no-change (between syncs) there won't be any network programming for neither of them and according the the SLI definition there is nothing to export. To give you more sophisticated example (E0,E1,E2... - endpoints states; A,B,C -> events) in case of:
(Sync) E0 -> A -> E1 -> B -> E2 -> C -> E0 -> D -> E3 -> E -> E4 (Sync)
the latency will be exported only for the trigger times of D and E.

Please correct me if you don't agree with my understanding / reasoning.

Reply:
TBH, I don't have very strong opinion which one is actually correct :-)
I'm fine with what you're proposing (i.e. skipping it in that case), but please add an explicit comment that this is a conscious decision.

Reply:
Done, added a comment.

Reply:
Okay. Noop sounds good.